web-console/src/views/query-view/query-input/query-input.tsx

```diff
@@ -25,45 +25,137 @@ import AceEditor from 'react-ace';
+          editor.completer.insertMatch({ value: data.caption });
+          const pos = editor.getCursorPosition();
+          editor.gotoLine(pos.row + 1, pos.column - 1);
+        },
+      },
+    };
+  });
+
+  langTools.addCompleter({
+    getCompletions: (_editor: any, _session: any, _pos: any, _prefix: any, callback: any) => {
+      callback(null, functionList);
+    },
+    getDocTooltip: (item: any) => {
+      if (item.meta === 'function') {
+        const functionName = item.caption.slice(0, -2);
+        item.docHTML = `
+<div class="function-doc">
+  <div class="function-doc-name">
+    <b>${escape(functionName)}</b>
+  </div>
+  <hr />
+  <div>
+    <b>Syntax:</b>
+  </div>
+  <div>${escape(item.syntax)}</div>
+  <br />
+  <div>
+    <b>Description:</b>
+  </div>
+  <div>${escape(item.description)}</div>
+</div>`;
+      }
+    },
+  });
+}
+
+function replaceDefaultAutoCompleter(): void {
+  if (!langTools) return;
+
+  const keywordList = ([] as any[]).concat(
+    SQL_KEYWORDS.map(v => ({ name: v, value: v, score: 0, meta: 'keyword' })),
+    SQL_EXPRESSION_PARTS.map(v => ({ name: v, value: v, score: 0, meta: 'keyword' })),
+    SQL_CONSTANTS.map(v => ({ name: v, value: v, score: 0, meta: 'constant' })),
+    SQL_DYNAMICS.map(v => ({ name: v, value: v, score: 0, meta: 'dynamic' })),
+    SQL_DATE_TYPES.map(v => ({ name: v.syntax, value: v.syntax, score: 0, meta: 'keyword' })),
+  );
+
+  const keywordCompleter = {
+    getCompletions: (_editor: any, _session: any, _pos: any, _prefix: any, callback: any) => {
+      return callback(null, keywordList);
+    },
+  };
+
+  langTools.setCompleters([langTools.snippetCompleter, langTools.textCompleter, keywordCompleter]);
+}
+
 export interface QueryInputProps {
   queryString: string;
   onQueryStringChange: (newQueryString: string) => void;
   runeMode: boolean;
   columnMetadata?: ColumnMetadata[];
+  currentSchema?: string;
+  currentTable?: string;
 }
 
 export interface QueryInputState {
   // For reasons (https://github.com/securingsincity/react-ace/issues/415) react ace editor needs an explicit height
   // Since this component will grown and shrink dynamically we will measure its height and then set it.
   editorHeight: number;
   prevColumnMetadata?: ColumnMetadata[];
+  prevCurrentTable?: string;
+  prevCurrentSchema?: string;
 }
 
 export class QueryInput extends React.PureComponent<QueryInputProps, QueryInputState> {
-  static getDerivedStateFromProps(props: ColumnTreeProps, state: ColumnTreeState) {
-    const { columnMetadata } = props;
+  static getDerivedStateFromProps(props: QueryInputProps, state: QueryInputState) {
+    const { columnMetadata, currentSchema, currentTable } = props;
+
+    if (
+      columnMetadata &&
+      (columnMetadata !== state.prevColumnMetadata ||
+        currentSchema !== state.prevCurrentSchema ||
+        currentTable !== state.prevCurrentTable)
+    ) {
+      replaceDefaultAutoCompleter();
+      addFunctionAutoCompleter();
```

Comment:
why is this being done on every prop change now?

Reply:
because if the table or schema changes I assumed we would also want to update the autocompleter. The default schema and table are coming from the ast that is only updated when you press run though.